src/injected/analyzers/tab-stops-orchestrator.ts

```diff
@@ -0,0 +1,104 @@
+    TabStopsRequirementEvaluator,
+} from 'injected/tab-stops-requirement-evaluator';
+import { TabbableElementGetter } from 'injected/tabbable-element-getter';
+import { FocusableElement } from 'tabbable';
+
+export class TabStopRequirementOrchestrator
+    implements AllFrameRunnerTarget<TabStopRequirementResult>
+{
+    public readonly name: string = 'TabStopRequirementOrchestrator';
+    private tabbableTabStops: FocusableElement[];
+    private actualTabStops: Set<HTMLElement> = new Set();
+    private latestVisitedTabStop: HTMLElement = null;
+    private reportResults: (payload: TabStopRequirementResult) => Promise<void>;
+
+    constructor(
+        private readonly dom: Document,
+        private readonly tabbableElementGetter: TabbableElementGetter,
+        private readonly windowUtils: WindowUtils,
+        private readonly tabStopsRequirementEvaluator: TabStopsRequirementEvaluator,
+        private readonly getUniqueSelector: (element: HTMLElement) => string,
+    ) {}
+
+    public start = () => {
+        this.dom.addEventListener('keydown', this.onKeydownForFocusTraps);
+        this.dom.addEventListener('focusin', this.addNewTabStop);
+        this.tabbableTabStops = this.tabbableElementGetter.getRawElements();
+        const tabbableFocusOrderResults =
+            this.tabStopsRequirementEvaluator.getTabbableFocusOrderResults(this.tabbableTabStops);
+        tabbableFocusOrderResults.forEach(result => {
+            this.reportResults(result);
+        });
+    };
+
+    public stop = () => {
+        this.dom.removeEventListener('keydown', this.onKeydownForFocusTraps);
+        this.dom.removeEventListener('focusin', this.addNewTabStop);
+        const keyboardNavigationResults =
+            this.tabStopsRequirementEvaluator.getKeyboardNavigationResults(
+                this.tabbableTabStops,
+                this.actualTabStops,
+            );
+        keyboardNavigationResults.forEach(this.reportResults);
+    };
+
+    public setResultCallback = (
+        reportResultCallback: (payload: TabStopRequirementResult) => Promise<void>,
+    ) => {
+        this.reportResults = reportResultCallback;
+    };
+
+    public transformChildResultForParent = (
+        result: TabStopRequirementResult,
+        messageSourceFrame: HTMLIFrameElement,
+    ): TabStopRequirementResult => {
+        const frameSelector = this.getUniqueSelector(messageSourceFrame);
+        result.selector = [frameSelector, ...result.selector];
+        return result;
+    };
+
+    private addNewTabStop = (focusEvent: FocusEvent) => {
+        const newTabStop = focusEvent.target as HTMLElement;
+        if (this.actualTabStops.size > 0 && this.actualTabStops.has(newTabStop)) {
+            return;
+        }
+
+        this.actualTabStops.add(newTabStop);
+        this.latestVisitedTabStop = newTabStop;
+
+        const result = this.tabStopsRequirementEvaluator.getFocusOrderResult(
+            this.latestVisitedTabStop,
+            newTabStop,
```

Comment:
Since you just assigned ` this.latestVisitedTabStop` on line 74, this calls `getFocusOrderResult` on the same tab stop for both curr and last, right? Do we need to move this call up a few lines before that assignment?

Reply:
Good catch. I was wondering myself but didn't go back to look into the notes myself..